File: ProjectTest/OneLang/CSharp/One/Transforms/InferTypesPlugins/ResolveFuncCalls.cs
using One.Transforms.InferTypesPlugins.Helpers;
using One.Ast;

namespace One.Transforms.InferTypesPlugins
{
    public class ResolveFuncCalls : InferTypesPlugin {
        public ResolveFuncCalls(): base("ResolveFuncCalls")
        {
            
        }
        
        public override bool canTransform(Expression expr) {
            return expr is UnresolvedCallExpression;
        }
        
        public override Expression transform(Expression expr) {
            var callExpr = ((UnresolvedCallExpression)expr);
            if (callExpr.func is GlobalFunctionReference globFunctRef) {
                var newExpr = new GlobalFunctionCallExpression(globFunctRef.decl, callExpr.args);
                callExpr.args = callExpr.args.map((Expression arg) => { return this.main.runPluginsOn(arg) ?? arg; });
                newExpr.setActualType(globFunctRef.decl.returns);
                return newExpr;
            }
            else {
                this.main.processExpression(expr);
                if (callExpr.func.actualType is LambdaType lambdType) {
                    var newExpr = new LambdaCallExpression(callExpr.func, callExpr.args);
                    newExpr.setActualType(lambdType.returnType);
                    return newExpr;
                }
                else
                    return expr;
            }
        }
    }
}

File: ProjectTest/OneLang/Java/src/main/java/OneLang/One/Transforms/InferTypesPlugins/ResolveFuncCalls/ResolveFuncCalls.java
import java.util.Arrays;

public class ResolveFuncCalls extends InferTypesPlugin {
    public ResolveFuncCalls()
    {
        super("ResolveFuncCalls");
        
    }
    
    public Boolean canTransform(Expression expr)
    {
        return expr instanceof UnresolvedCallExpression;
    }
    
    public Expression transform(Expression expr)
    {
        var callExpr = ((UnresolvedCallExpression)expr);
        if (callExpr.func instanceof GlobalFunctionReference) {
            var newExpr = new GlobalFunctionCallExpression(((GlobalFunctionReference)callExpr.func).decl, callExpr.args);
            callExpr.args = Arrays.stream(callExpr.args).map(arg -> this.main.runPluginsOn(arg) != null ? this.main.runPluginsOn(arg) : arg).toArray(Expression[]::new);
            newExpr.setActualType(((GlobalFunctionReference)callExpr.func).decl.returns);
            return newExpr;
        }
        else {
            this.main.processExpression(expr);
            if (callExpr.func.actualType instanceof LambdaType) {
                var newExpr = new LambdaCallExpression(callExpr.func, callExpr.args);
                newExpr.setActualType(((LambdaType)callExpr.func.actualType).returnType);
                return newExpr;
            }
            else
                return expr;
        }
    }
}